vnext/JSI/Shared/ChakraObjectRef.cpp

```diff
@@ -0,0 +1,325 @@
+  }
+}
+
+ChakraObjectRef::ChakraObjectRef(const ChakraObjectRef &original) noexcept
+    : m_ref{original.m_ref}, m_state{original.m_state} {
+  if (m_state == State::Initialized) {
+    assert(m_ref);
+    CheckedJsAddRef(m_ref);
+  } else {
+    assert(!m_ref);
+  }
+}
+
+ChakraObjectRef::ChakraObjectRef(ChakraObjectRef &&original) noexcept {
+  swap(original);
+}
+
+ChakraObjectRef &ChakraObjectRef::operator=(
+    const ChakraObjectRef &rhs) noexcept {
+  ChakraObjectRef rhsCopy(rhs);
+  swap(rhsCopy);
+  return *this;
+}
+
+ChakraObjectRef &ChakraObjectRef::operator=(ChakraObjectRef &&rhs) noexcept {
+  swap(rhs);
+  return *this;
+}
+
+ChakraObjectRef::~ChakraObjectRef() noexcept {
+  if (m_state == State::Initialized) {
+    assert(m_ref);
+    CheckedJsRelease(m_ref);
+  } else {
+    assert(!m_ref);
+  }
+}
+
+void ChakraObjectRef::Initialize(JsRef ref) {
+  assert(!m_ref);
+
+  if (m_state != State::Uninitialized) {
+    throw facebook::jsi::JSINativeException(
+        "A ChakraObjectRef can only be initialzed once.");
+  }
+
+  if (!ref) {
+    throw facebook::jsi::JSINativeException(
+        "Cannot initialize a ChakraObjectRef with a null reference.");
+  }
+
+  CheckedJsAddRef(ref);
+  m_ref = ref;
+  m_state = State::Initialized;
+}
+
+void ChakraObjectRef::Invalidate() {
+  switch (m_state) {
+    case State::Uninitialized: {
+      assert(!m_ref);
+      throw facebook::jsi::JSINativeException(
+          "Cannot invalidate a ChakraObjectRef that has not been initialized.");
+      break;
+    }
+    case State::Initialized: {
+      assert(m_ref);
+      CheckedJsRelease(m_ref);
+      m_ref = nullptr;
+      m_state = State::Invalidated;
+      break;
+    }
+    case State::Invalidated: {
+      assert(!m_ref);
+      throw facebook::jsi::JSINativeException(
+          "Cannot invalidate a ChakraObjectRef that has already been "
+          "invalidated.");
+      break;
+    }
+    default: {
+      // Control flow should never reach here.
+      std::terminate();
+      break;
+    }
+  }
+}
+
+void ChakraObjectRef::swap(ChakraObjectRef &other) {
+  std::swap(m_ref, other.m_ref);
+  std::swap(m_state, other.m_state);
+}
+
+JsValueType GetValueType(const ChakraObjectRef &jsValue) {
+  JsValueType type;
+  ThrowUponChakraError(JsGetValueType(jsValue, &type), "JsGetValueType");
+  return type;
+}
+
+JsPropertyIdType GetPropertyIdType(const ChakraObjectRef &jsPropId) {
+  JsPropertyIdType type;
+  ThrowUponChakraError(
+      JsGetPropertyIdType(jsPropId, &type), "JsGetPropertyIdType");
+  return type;
+}
+
+std::wstring GetPropertyName(const ChakraObjectRef &id) {
+  if (GetPropertyIdType(id) != JsPropertyIdTypeString) {
+    throw facebook::jsi::JSINativeException(
+        "It is llegal to retrieve the name of a property symbol.");
+  }
+  const wchar_t *propertyName = nullptr;
+  ThrowUponChakraError(
+      JsGetPropertyNameFromId(id, &propertyName), "JsGetPropertyNameFromId");
+  return std::wstring{propertyName};
```

Comment:
>propertyName [](start = 22, length = 12)

Who is responsible for cleaning this string? Or it's alive while Id is alive?

Reply:
This buffer is alive as long as the Chakra runtime is alive. (See [https://github.com/microsoft/ChakraCore/wiki/JsGetPropertyNameFromId](https://github.com/microsoft/ChakraCore/wiki/JsGetPropertyNameFromId).)

---
In reply to: [336173945](https://github.com/microsoft/react-native-windows/pull/3445#discussion_r336173945) [](ancestors = 336173945)